### PR TITLE
Add a playbook for installing and configuring Graphite.

### PR DESCRIPTION
**Pull request by tobz**
This playbook will install `carbon-cache` and `graphite-api` to provide a minimalistic starting point for metrics collection.  A server would be ready to accept metrics (from anything that could speak the Graphite line protocol, such as `statsd`) and be a target for a Grafana installation after running this playbook on its own.

It currently installs all components from their upstream source (albeit with defaults that point to tags, and not master/development branches).  It uses systemd, and so is tailored specifically to Ubuntu 16.04.

There is a small "master" playbook -- `edx-monitoring` -- for running the role, and in the future, supporting the culmination of other playbooks to build out an all-in-one monitoring server: Graphite, Grafana, etc.

Configuration Pull Request
---

Make sure that the following steps are done before merging

  - [x] A devops team member has commented with :+1:
  - [ ] are you adding any new default values that need to be overridden when this goes live?
    - [ ] Open a ticket (DEVOPS) to make sure that they have been added to secure vars.
    - [ ] Add an entry to the CHANGELOG.

### CODE REVIEW

**Comment by tobz** [comment]
Point of order: I cribbed this playbook from an existing playbook (https://github.com/azavea/ansible-graphite) and so the LICENSE from the original project is included, with an amendment at the top pointing back to the original project.

This should be kosher given that it's the Apache 2.0 license.

**Comment by e0d** [comment]
@tobz high-level question, can you summarize what changes you made from the source ansible artifacts and why you chose this one?

**Comment by tobz** [comment]
High-level changes:
- make it use a virtualenv
- install graphite-api over graphite-web (don't need all the web baggage if running Grafana on top)
- make it install gunicorn (because we're using nginx, so no mod_wsgi)
- use systemd (instead of SysV) to ensure both carbon-cache and graphite-api get supervised properly

I mostly chose this one because it worked off of the tagged source, had a suitable license for derivations, and was one of the smaller ones that I found.  Many of them were versions that appeared to be tailored to the individual's own needs / company needs.

File: playbooks/roles/graphite/handlers/main.yml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+---
+- name: Restart Carbon
```
Comment (e0d) [inline playbooks/roles/graphite/handlers/main.yml]: We've been typically avoiding handlers as their behavior can be surprising and we typically don't do rolling upgrades.  We've been using tags to target tasks to give us flexibility around which sets are run at any invocation:

https://openedx.atlassian.net/wiki/display/OpenOPS/Role+Life-cycle+Tags

File: playbooks/roles/graphite/tasks/carbon.yml [hunk 2]
```diff
@@ -0,0 +1,49 @@
+---
+- name: checkout Carbon
+  git: repo=https://github.com/graphite-project/carbon.git
+       dest={{ graphite_root }}/src/carbon
+       version={{ GRAPHITE_CARBON_VERSION }}
+
+- name: install Carbon dependencies
```
Comment (e0d) [inline playbooks/roles/graphite/tasks/carbon.yml]: why not use the pip module?

File: playbooks/roles/graphite/tasks/carbon.yml [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+            dest={{ graphite_root }}/conf/{{ item }}.conf
+  with_items:
+    - carbon
+    - storage-schemas
+    - storage-aggregation
+  notify:
+    - Restart Carbon
+
+- name: set permissions for storage
+  file: path={{ graphite_root }}/storage
```
Comment (e0d) [inline playbooks/roles/graphite/tasks/carbon.yml]: Feels like we should separate out the storage root from the app root.  It may be that by default graphite_storage is set to {{ graphite_root }}/storage, but I'd think we'd relocate the storage at scale.

File: playbooks/roles/graphite/tasks/graphite-api.yml [hunk 4]
```diff
@@ -0,0 +1,49 @@
+---
+- name: checkout Graphite API
+  git: repo=https://github.com/brutasse/graphite-api.git
```
Comment (e0d) [inline playbooks/roles/graphite/tasks/graphite-api.yml]: repo should be a var I think, makes groking from defaults easier, supports forks, etc.

File: playbooks/roles/graphite/tasks/graphite-api.yml [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+       dest={{ graphite_root }}/src/api
+       version={{ GRAPHITE_API_VERSION }}
+
+- name: install Graphite API dependencies
+  command: "{{ graphite_root }}/bin/pip install -r {{ graphite_root }}/src/api/requirements.txt"
```
Comment (e0d) [inline playbooks/roles/graphite/tasks/graphite-api.yml]: pip module

File: playbooks/roles/graphite/tasks/main.yml [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+- name: create service account for Carbon
+  user: name=carbon
+        system=yes
+        home={{ graphite_root }}
+        shell=/bin/false
+        group=graphite
+        state=present
+
+- name: create common Graphite directories
```
Comment (e0d) [inline playbooks/roles/graphite/tasks/main.yml]: I like doing this using with_items and data:

https://github.com/edx/configuration/blob/master/playbooks/roles/vhost/defaults/main.yml#L18

File: playbooks/roles/graphite/tasks/main.yml [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+- name: create Graphite source directories
+  file: path={{ graphite_root }}/src
+        state=directory
+        owner=graphite
+        group=graphite
+
+- name: install Graphite dependencies
+  apt: pkg={{ item }}
+       state=present
+  with_items:
```
Comment (e0d) [inline playbooks/roles/graphite/tasks/main.yml]: use a var: https://github.com/edx/configuration/blob/master/playbooks/roles/common_vars/defaults/main.yml#L110

File: playbooks/roles/graphite/defaults/main.yml [hunk 8]
```diff
@@ -0,0 +1,34 @@
+---
```
Comment (e0d) [inline playbooks/roles/graphite/defaults/main.yml:1]: Can you add boilerplate and explanatory comment, c.f., https://github.com/edx/configuration/blob/master/playbooks/roles/vhost/tasks/main.yml#L2

File: playbooks/roles/graphite/templates/carbon/conf/carbon.conf.j2 [hunk 9]
```diff
@@ -0,0 +1,23 @@
+[cache]
+USER = carbon
+CACHE_WRITE_STRATEGY = max
+MAX_CACHE_SIZE = inf
+USE_FLOW_CONTROL = True
+WHISPER_FALLOCATE_CREATE = True
+MAX_CREATES_PER_MINUTE = 100
+MAX_UPDATES_PER_SECOND = 1000
+LINE_RECEIVER_INTERFACE = 0.0.0.0
```
Comment (e0d) [inline playbooks/roles/graphite/templates/carbon/conf/carbon.conf.j2]: Seems like some of these are things we'd want to var out.  Listening on 0.0.0.0 for sure seems like something we'd want to make obvious in the defaults.  Other things we'd expect ourselves or others to reasonably change should be as well.

File: playbooks/roles/nginx/templates/edx/app/nginx/sites-available/graphite.j2 [hunk 10]
```diff
@@ -0,0 +1,52 @@
+#
```
Comment (e0d) [inline playbooks/roles/nginx/templates/edx/app/nginx/sites-available/graphite.j2:1]: Play will need to include nginx.

File: playbooks/roles/graphite/templates/graphite-api/systemd/graphite-api.service.j2 [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+Requires=graphite-api.socket
+
+[Service]
+StandardOutput=syslog
+StandardError=syslog
+User=graphite
+Group=graphite
+WorkingDirectory={{ graphite_root }}
+Environment=GRAPHITE_API_CONFIG={{ graphite_root }}/conf/graphite-api.yaml
+ExecStart={{ graphite_root }}/bin/gunicorn -w 2 graphite_api.app:app
```
Comment (e0d) [inline playbooks/roles/graphite/templates/graphite-api/systemd/graphite-api.service.j2]: I think workers should be configurable.

File: playbooks/roles/graphite/tasks/main.yml [hunk 12]
```diff
@@ -0,0 +1,68 @@
+---
+- name: create Graphite user group
+  group: name=graphite
```
Comment (e0d) [inline playbooks/roles/graphite/tasks/main.yml]: we should make this a var and the group so we can pass to templates later.

File: playbooks/roles/graphite/templates/graphite-api/systemd/graphite-api.service.j2 [hunk 13]
```diff
@@ -0,0 +1,20 @@
+[Unit]
+Description=Graphite-API service
+Requires=graphite-api.socket
+
+[Service]
+StandardOutput=syslog
+StandardError=syslog
+User=graphite
```
Comment (e0d) [inline playbooks/roles/graphite/templates/graphite-api/systemd/graphite-api.service.j2]: Use user/group vars defined above.

File: playbooks/roles/graphite/tasks/main.yml [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+- name: initialize virtualenv for Graphite
+  command: "virtualenv {{ graphite_root }}"
+  args:
+    creates: "{{ graphite_root }}/bin/activate"
+
+- include: whisper.yml
+- include: carbon.yml
+- include: graphite-api.yml
+
+- name: Find all example configuration files
```
Comment (e0d) [inline playbooks/roles/graphite/tasks/main.yml]: what's wrong with the examples?

**Comment by e0d** [comment]
I think we should have a docker CI build that can run in travis:

https://github.com/edx/configuration/tree/master/docker/build

**Review by e0d — COMMENTED** [review]
First pass.

File: playbooks/roles/graphite/tasks/main.yml (hunk 14 above)
Comment (tobz) [reply playbooks/roles/graphite/tasks/main.yml]: Nothing, really.  There's just like 10 of them and it felt like it was just a little harder than it ought to be to peek at the file listing and see the actual configuration files being used.

File: playbooks/roles/graphite/tasks/carbon.yml (hunk 2 above)
Comment (tobz) [reply playbooks/roles/graphite/tasks/carbon.yml]: There's a pip module? Sweet! I'll use that.

File: playbooks/roles/graphite/tasks/main.yml (hunk 12 above)
Comment (tobz) [reply playbooks/roles/graphite/tasks/main.yml]: Good point.  Will do.

File: playbooks/roles/graphite/defaults/main.yml (hunk 8 above)
Comment (tobz) [reply playbooks/roles/graphite/defaults/main.yml:1]: I addressed this in `tasks/main.yml` specifically.

File: playbooks/roles/nginx/templates/edx/app/nginx/sites-available/graphite.j2 (hunk 10 above)
Comment (tobz) [reply playbooks/roles/nginx/templates/edx/app/nginx/sites-available/graphite.j2:1]: Yup, this is now taken care of.

File: playbooks/roles/graphite/handlers/main.yml (hunk 1 above)
Comment (tobz) [reply playbooks/roles/graphite/handlers/main.yml]: I opted to stick with handlers for the time being.

**Comment by tobz** [comment]
@edx/devops Since Ed is out, can someone take another pass at this for review?

**Comment by fredsmith** [comment]
@tobz can you add tags on all tasks according to https://openedx.atlassian.net/wiki/display/OpenOPS/Role+Life-cycle+Tags ?  We've been doing this for all ansible code that gets merged.

You should also re-write your handlers as tasks, and make sure they're tagged appropriately.

**Comment by tobz** [comment]
@fredsmith This should be all set now.

File: playbooks/roles/graphite/handlers/main.yml (hunk 1 above)
Comment (fredsmith) [reply playbooks/roles/graphite/handlers/main.yml]: I think this can go away now.

**Review by fredsmith — CHANGES_REQUESTED** [review]
I think we can remove the handler definitions, then :+1:

**Comment by tobz** [comment]
Ah right, my mistake.  Fixed.

**Review by fredsmith — APPROVED** [review]

**Comment by fredsmith** [comment]
looks good to merge :+1:
